enhancements/storage/csi-snapshot.md

```diff
@@ -37,52 +38,98 @@ We want to include the upstream CSI Snapshot feature in OpenShift.
+* Enable snapshot feature by default and provide the same support as upstream 1.17 release.
 
-* Package and ship a downstream image of the this sidecar.
+* Release and maintain downstream csi-external-snapshotter, csi-external-provisioner sidecars images off of the upstream images based on Kubernetes 1.17 to allow Red Hat CSI drivers (such as OCS) to consume them.
 
 ### Non-Goals
 
+## Upstream status
+This feature has reached v1beta status upstream and is described in https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190709-csi-snapshot.md and https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-snapshot.md.
+
+This feature poses new requirements on Kubernetes distributions such as OpenShift:
+* Ship and install snapshot CRDs.
+* Ship and run snapshot-controller in a cluster.
+  * snapshot-controller is very similar to in-tree PV controller in its function, it watches VolumeSnapshots and provisions VolumeSnapshotContents for them. The snapshot controller is independent on storage backend and only one runs for all CSI drivers installed in a cluster.
+
+Upstream has chosen [addon](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/volumesnapshots) for both. Upstream uses [StatefulSet for the snapshot-controller with single replica](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/volumesnapshots/volume-snapshot-controller/volume-snapshot-controller-deployment.yaml) for the snapshot-controller.
+
 ## Proposal
 
-This feature has already been implemented upstream, and is described in https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190709-csi-snapshot.md . It is currently documented upstream at https://kubernetes.io/docs/concepts/storage/volume-snapshots/ .
+1. OCP ships a new operator for csi-snapshot-controller, say csi-snapshot-controller-operator.
+  * This operator will be installed by CVO in all clusters.
+    * We do not know in advance what CSI drivers will a cluster admin install and we want the cluster ready for snapshots.
+  * This operator will create Deployment with csi-snapshot-controller and report its status via standard `ClusterOperator` object.
+    * Using Deployment with leader election instead of upstream StatefulSet - leader election is significantly faster to run a new leader when a node with the current leader gets unavailable.
+    * TBD: Use DaemonSet on masters? The controller can read / write any VolumeSnapshot, VolumeSnapshotContent and PV objects in the cluster and read any PVC. Running it on masters may be more secure.
+  * Everything runs in namespace `openshift-csi-snapshot-controller`.
+  * Requires new github repo, openshift/csi-snapshot-controller-operator.
+  * Requires new image.
+  * No new CRD / CR is needed for the operator itself, the operator does the same on all clusters / clouds and does not need any configuration.
```

Comment:
We've found that when building up clusteroperator status it can be useful to have individual FooDegraded and FooAvailable conditions in a myoperator.operator.openshift.io resource and aggregate those conditions for the clusteroperator status.  Persisting the individual conditions makes it very easy to add fully independent loops and allows detailed individual methods.  We use this technique in group-b operators to good effect and have library-go control loops to make the individual wiring easier.

Reply:
It is worth in this operator that starts just installs CRDs and in the end runs a simple Deployment?

Reply:
> It is worth in this operator that starts just installs CRDs and in the end runs a simple Deployment?

I think so.  It turns out that most operators are impacted by running conditions in the rest of the cluster that are not directly related and the ability to add sanity checks for various degraded condition produces a lot of value by focusing how bugs are redirected and how quickly.  You can imagine sanity checks being built for all sorts of conditions including potentially health of the plugged in units or alerts on too many failures in status for your CRs.   I think the flexibility adds great value for little cost.

group-b even has history adding conditions to operators in other groups.